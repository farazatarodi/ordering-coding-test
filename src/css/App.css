/* global variables and component styles are stored here */

.App {
  width: 100%;
  height: 100%;
  display: flex;

  --shadow: 0 0 10px 0px rgba(0, 0, 0, 0.07);
  --shadow-hover: 0 0 18px 0px rgba(0, 0, 0, 0.4);
  --radius: 1rem;
}

.content {
  position: relative;
  width: 80%;
  height: 100%;
  padding: 4rem;
  background-color: var(--color-neutral-light);
  display: flex;
  justify-content: center;
  overflow: auto;
}

/* style for card components */
.card {
  width: 100%;
  height: max-content;
  background-color: var(--color-neutral-lightest);
  border-top: 2px solid var(--color-neutral);
  padding: 2rem;
  display: grid;
  grid-template-columns: 10% auto 20% 20%;
  align-items: center;
}

.card:last-child {
  border-radius: 0 0 var(--radius) var(--radius);
}

/* style for buttons */
.btn {
  border: none;
  padding: 0.5rem;
  border-radius: calc(var(--radius) / 2);
  color: var(--color-neutral-lightest);
  background-color: var(--color-mint);
  cursor: pointer;
  transition: all 0.1s ease-in-out;
}

.btn:hover {
  background-color: var(--color-mint-dark);
}
