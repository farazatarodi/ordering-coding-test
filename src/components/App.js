// React imports
import { useEffect } from 'react';

// Router imports
import { Route, BrowserRouter as Router, Switch } from 'react-router-dom';

// Redux imports
import { useDispatch } from 'react-redux';
import { fetchCustomers } from '../redux/actions/customers';
import { fetchOrders } from '../redux/actions/orders';
import { fetchProducts } from '../redux/actions/products';

// component imports
import Customers from './Customers';
import Orders from './Orders';
import Products from './Products';
import Sidebar from './Sidebar';

// css imports
import '../css/App.css';

function App() {
  // prepare dispatch for callback use
  const dispatch = useDispatch();

  // fetch customers, orders and products when the app loads
  useEffect(() => {
    dispatch(fetchCustomers());
    dispatch(fetchOrders());
    dispatch(fetchProducts());
  }, [dispatch]);

  return (
    <div className="App">
      <Router>
        <Sidebar />
        <div className="content">
          {/* main switch only changes the content section */}
          <Switch>
            {/* Orders page */}
            <Route path="/orders">
              <Orders />
            </Route>

            {/* Products page */}
            <Route path="/products">
              <Products />
            </Route>

            {/* Customers page (homepage) */}
            <Route path="/">
              <Customers />
            </Route>
          </Switch>
        </div>
      </Router>
    </div>
  );
}

export default App;
