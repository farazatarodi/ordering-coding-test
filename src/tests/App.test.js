import { render, screen, fireEvent } from '@testing-library/react';

import App from '../components/App';

import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';

import renderer from 'react-test-renderer';

test('on app open we see sidebar and customers page', () => {
  const initialState = { customersReducer: { customers: [] } };
  const mockStore = configureStore();
  const store = mockStore(initialState);

  render(
    <Provider store={store}>
      <App />
    </Provider>
  );

  expect(screen.getAllByText(/customers/i)).toHaveLength(2);
});

test('we can navigate through different pages', () => {
  const initialState = {
    customersReducer: { customers: [] },
    ordersReducer: { orders: [] },
    productsReducer: { products: [] },
  };
  const mockStore = configureStore();
  const store = mockStore(initialState);

  render(
    <Provider store={store}>
      <App />
    </Provider>
  );

  fireEvent.click(screen.getByRole('link', { name: /orders/i }));
  expect(screen.getAllByText(/orders/i)).toHaveLength(2);

  fireEvent.click(screen.getByRole('link', { name: /products/i }));
  expect(screen.getAllByText(/products/i)).toHaveLength(2);
});

test('renders correclty', () => {
  const initialState = {
    customersReducer: { customers: [] },
    ordersReducer: { orders: [] },
    productsReducer: { products: [] },
  };
  const mockStore = configureStore();
  const store = mockStore(initialState);
  const tree = renderer
    .create(
      <Provider store={store}>
        <App />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
