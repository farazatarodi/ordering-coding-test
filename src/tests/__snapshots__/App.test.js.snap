// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correclty 1`] = `
<div
  className="App"
>
  <div
    className="sidebar"
  >
    <img
      alt="Teamleader Logo"
      src="logo.svg"
    />
    <a
      className="side-item heading-1 heading"
      href="/"
      onClick={[Function]}
    >
      <svg
        className="side-item-icon"
        fill="currentColor"
        height="24"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "color": undefined,
            "opacity": 0.84,
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M18 20c0-3.309-2.691-6-6-6s-6 2.691-6 6H4c0-4.411 3.589-8 8-8s8 3.589 8 8h-2zM4 20h2v1a1 1 0 01-2 0v-1zm14 0h2v1a1 1 0 01-2 0v-1zm-6-10a4 4 0 110-8 4 4 0 010 8zm0-2a2 2 0 100-4 2 2 0 000 4z"
          fillRule="evenodd"
        />
      </svg>
      Customers
    </a>
    <a
      className="side-item heading-1 heading"
      href="/orders"
      onClick={[Function]}
    >
      <svg
        className="side-item-icon"
        fill="currentColor"
        height="24"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "color": undefined,
            "opacity": 0.84,
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          clipRule="evenodd"
          d="M9 6H6v3h3V6zM4 4v7h7V4H4zm5 11H6v3h3v-3zm-5-2v7h7v-7H4zm14-7h-3v3h3V6zm-5-2v7h7V4h-7zm5 11h-3v3h3v-3zm-5-2v7h7v-7h-7z"
          fillRule="evenodd"
        />
      </svg>
      Orders
    </a>
    <a
      className="side-item heading-1 heading"
      href="/products"
      onClick={[Function]}
    >
      <svg
        className="side-item-icon"
        fill="currentColor"
        height="24"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "color": undefined,
            "opacity": 0.84,
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M5 5v8h4V5H5zm6 4.9V13h3.1L11 9.9zm0-3.314l3.293-3.293a1 1 0 011.414 0l4.243 4.243a1 1 0 010 1.414L15.9 13H19a2 2 0 012 2v4a2 2 0 01-2 2H5a2 2 0 01-2-2v-4c0-.364.097-.706.268-1A1.99 1.99 0 013 13V5a2 2 0 012-2h4a2 2 0 012 2v1.586zM5 15v4h14v-4H5zm7.172-6.757L15 11.07l2.828-2.828L15 5.414l-2.828 2.829z"
        />
      </svg>
      Products
    </a>
  </div>
  <div
    className="content"
  >
    <div
      className="products-container"
    >
      <h1
        className="heading heading-1"
      >
        Products
      </h1>
      <div
        className="products-content"
      >
        <div
          className="products-header heading heading-2"
        >
          <strong>
            ID
          </strong>
          <strong>
            Description
          </strong>
          <strong>
            Category
          </strong>
          <strong>
            Price
          </strong>
          <strong>
            Add to order
          </strong>
        </div>
        <div
          className="card text text-display notification-card"
        >
          Nothing to show
        </div>
      </div>
    </div>
  </div>
</div>
`;
